src/webgpu/api/validation/encoding/cmds/setBindGroup.spec.ts

```diff
@@ -4,113 +4,137 @@ setBindGroup validation tests.
+        unreachable('unknown resource type');
+    }
   }
 
-  testRenderPass(bindGroup: GPUBindGroup, dynamicOffsets: number[]): void {
-    const encoder = this.device.createCommandEncoder();
-    const renderPass = encoder.beginRenderPass({
-      colorAttachments: [
-        {
-          attachment: this.makeAttachmentTexture().createView(),
-          loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
-        },
-      ],
+  createBindGroup(
+    state: 'valid' | 'invalid' | 'destroyed',
+    resourceType: 'buffer' | 'texture',
+    encoderType: ProgrammableEncoderType,
+    indices: number[]
+  ) {
+    const layoutIndices =
+      state === 'invalid' ? new Array<number>(indices.length + 1).fill(0) : indices;
+    const layout = this.device.createBindGroupLayout({
+      entries: layoutIndices.map(binding => ({
+        binding,
+        visibility: this.encoderTypeToStageFlag(encoderType),
+        ...(resourceType === 'buffer'
+          ? {
+              buffer: {
+                type: 'storage',
+              },
+            }
+          : { texture: {} }),
+      })),
     });
-    renderPass.setBindGroup(0, bindGroup, dynamicOffsets);
-    renderPass.endPass();
-    encoder.finish();
-  }
-
-  testRenderBundle(bindGroup: GPUBindGroup, dynamicOffsets: number[]): void {
-    const encoder = this.device.createRenderBundleEncoder({
-      colorFormats: ['rgba8unorm'],
+    return this.device.createBindGroup({
+      layout,
+      entries: indices.map(binding => ({
+        binding,
+        resource: this.createResourceWithState(
+          resourceType,
+          state === 'destroyed' ? state : 'valid'
+        ),
+      })),
     });
-    encoder.setBindGroup(0, bindGroup, dynamicOffsets);
-    encoder.finish();
   }
 }
 
 export const g = makeTestGroup(F);
 
-g.test('dynamic_offsets_passed_but_not_expected,compute_pass')
-  .params(poptions('type', ['compute', 'renderpass', 'renderbundle']))
-  .fn(async t => {
-    const bindGroupLayout = t.device.createBindGroupLayout({ entries: [] });
-    const bindGroup = t.device.createBindGroup({ layout: bindGroupLayout, entries: [] });
+g.test('state_and_binding_index')
+  .desc('Tests that setBindGroup correctly handles {valid, invalid} bindGroups.')
+  .cases(
+    params()
+      .combine(poptions('encoderType', kProgrammableEncoderTypes))
+      .combine(poptions('state', ['valid', 'invalid', 'destroyed'] as const))
+      .combine(poptions('resourceType', ['buffer', 'texture'] as const))
+  )
+  .fn(t => {
+    const { encoderType, state, resourceType } = t.params;
+    const { maxBindGroups } = t.device.adapter.limits || { maxBindGroups: 4 };
+
+    function runTest(index: number) {
+      const { encoder, finish } = t.createEncoder(encoderType);
+
+      t.expectValidationError(() => {
+        encoder.setBindGroup(index, t.createBindGroup(state, resourceType, encoderType, [index]));
+        // Submit required to test destroyed buffer
+        t.queue.submit([finish()]);
```

Comment:
Good point on being unable to tell what caused the error. Have you considered, instead of exposing more details to the test, just moving the expectValidationError checks to `finish`? and/or creating another expectValidationError-like function that understands CommandBufferMaker? Otherwise, it seems like each test that uses createEncoder would gain extra boilerplate.

Regarding `api/validation/queue/destroyed/*`, I'm not a huge fan of having similar test logic in 2 places. Maybe we can add a method to `ValidationTest` that conditionally checks either finish or submit depending on a state?
```ts
  expectValidationErrorWithState(
    state: 'valid' | 'invalid' | 'destroyed',
    finish: () => Iterable<GPUCommandBuffer>,
    shouldError: boolean = true
  ): void {
    if (state === 'destroyed') {
      const commandBuffers = finish();
      this.expectValidationError(() => {
        this.queue.submit(commandBuffers);
      }, shouldError);
    } else {
      this.expectValidationError(() => {
        finish();
      }, shouldError);
    }
  }
```

Reply:
`shouldError` is not required but useful. Even the current test adds `index >= maxBindGroups`. Might be nice to also incorporate the state check into the function.
```ts
  expectValidationErrorWithState(
    state: 'valid' | 'invalid' | 'destroyed',
    finish: () => Iterable<GPUCommandBuffer>,
    shouldError: boolean = true
  ): void {
    if (state === 'destroyed') {
      const commandBuffers = finish();
      this.expectValidationError(() => {
        this.queue.submit(commandBuffers);
      });
    } else {
      this.expectValidationError(() => {
        finish();
      }, state !== 'valid' || shouldError);
    }
  }
```

Reply:
Ugh, I just noticed your point. The problem with destroy test is that if finish() fails, the test will fail. I guess we could either
1. Wrap finish with another validation error scope. This will cause an "Object is an error." instead of "Destroyed buffer used in a submit" error.
2. Keep `expectValidationErrorWithState` as is and force the tests to making sure it doesn't happens. Maybe add `assert(shouldError, 'destroyed must always fail.');` to make the error is more clear?
3. Remove `shouldError`, split the code into 2 tests. Your original suggestion.

I prefer 1. While it adds extra test cases and hides true reason for the failure, it's the only choice that doesn't add extra code to the tests themselves.

What do you think?

Reply:
To clarify, `finish()` must succeed if the resource was originally valid but then destroyed, and we want to test that. I think 1 sounds good to me.

Reply:
I'll split this into a separate PR